lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      continue;
     }
-    case ElementInfo::Ground: {
-      auto type = builder.getIntegerType(info.width);
-      builder.create<sv::InterfaceSignalOp>(loc, name, type);
-      break;
+    if (companionIDMap.count(bundle.getID()) == 0) {
+      emitCircuitError() << "no companion found with 'id' value '"
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      continue;
     }
-    case ElementInfo::String:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported string type>;");
-      break;
-    case ElementInfo::Boolean:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported boolean type>;");
-      break;
-    case ElementInfo::Integer:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported integer type>;");
-      break;
-    case ElementInfo::Double:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported double type>;");
-      break;
-    case ElementInfo::Error:
-      llvm_unreachable("Shouldn't be here");
-      break;
+
+    // Recursively walk the AugmentedBundleType to generate interfaces and XMRs.
+    // Error out if this returns None (indicating that the annotation annotation
+    // is malformed in some way).  A good error message is generated inside
+    // `traverseBundle` or the functions it calls.
+    auto iface = traverseBundle(bundle, bundle.getID(),
+                                companionIDMap.lookup(bundle.getID()).name);
+    if (!iface) {
+      removalError = true;
+      continue;
     }
+
+    // Instantiate the interface inside the parent.
+    builder.setInsertionPointToEnd(
+        parentIDMap.lookup(bundle.getID()).second.getBodyBlock());
+    auto symbolName = getNamespace().newName(
+        ("__" + companionIDMap.lookup(bundle.getID()).name + "_" +
+         bundle.getDefName().getValue() + "__"));
+    auto instance = builder.create<sv::InterfaceInstanceOp>(
+        getOperation().getLoc(), iface.getValue().getInterfaceType(),
+        companionIDMap.lookup(bundle.getID()).name,
+        builder.getStringAttr(symbolName));
+
+    // If no extraction information was present, then just leave the interface
+    // instantiated in the parent.  Otherwise, make it a bind.
+    if (!maybeExtractInfo)
+      continue;
+
+    instance->setAttr("doNotPrint", builder.getBoolAttr(true));
+    builder.setInsertionPointToStart(
+        instance->getParentOfType<ModuleOp>().getBody());
+    auto bind = builder.create<sv::BindInterfaceOp>(
+        getOperation().getLoc(),
+        SymbolRefAttr::get(builder.getContext(),
+                           instance.sym_name().getValue()));
```

Comment:
Does this work?
```c++
        SymbolRefAttr::get(instance.sym_name()));
```

Reply:
It does not. 😢

Reply:
Maybe its `SymbolRefAttr::get(instance.sym_nameAttr()));`

Reply:
Looks like both methods return `::mlir::StringAttr`, I thought there was a cheaper way to go from `StringAtttr` -> `SymbolRef` instead of jumping through StringRef.